### PR TITLE
Support user-defined bin edges in binned lightcurves

### LINKED ISSUE #628: Support for passing in list of bin placements when binning a lightcurve.
Imagine I want to bin a lightcurve into a fixed number of points, or I want to bin a lightcurve at variable-width intervals with known boundaries in a list.  The current `.bin()` method currently accepts a `binsize=` parameter, which is the number of cadences to include in every bin.  

We could add support for these time-based bins using functions similar to those in astropy's histogram features, e.g. `bin_edges`

### PR DESCRIPTION
**Pull request by gully**
This PR will close #628.

### CODE REVIEW

File: lightkurve/tests/test_lightcurve.py [hunk 1]
```diff
@@ -435,6 +435,67 @@ def test_bin():
...
+    assert len(binned_lc) == n_times // 10
+    binned_lc = lc.bin(binsize=11)
+    # Resulting length with binsize may depend on implementation:
+    #   Allowing for under-filled bins at boundary conditions
+    assert ((len(binned_lc) >= (n_times // 11) ) &
+            (len(binned_lc) <= (n_times // 11 +1) ) )
+    #   Not allowing for under-filled bins at boundary conditions
+    assert (len(binned_lc) == (n_times // 11)
+    # Resulting length with `bins=N` yields exactly N bins every time
+    binned_lc == lc.bin(bins=38)
```
Comment (barentsen) [inline lightkurve/tests/test_lightcurve.py]: Codacy Issue found: [invalid syntax](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4502630)

**Comment by gully** [comment]
Next step: replace the heritage code block with a streamlined code with `bin_by=self.time` or `bin_by=np.arange(0, len(self.time))`, and update the bin_edges and n_bins accordingly.

**Comment by gully** [comment]
Question: 
**Should we allow zero values for `flux_err` in binned lightcurves?**

Background:
`.bin()` falls back to computing the binned flux_err as the standard deviation of flux values in the bin.  If the lightcurve has either one point in a bin, or exactly equal values in a bin, the standard deviation is zero.  A zero standard deviation is, well, a zero uncertainty, which is unphysical and cause problems elsewhere.

Proposed solution:
We should have a `validate_flux_err()` utility somewhere that looks for zero (and possibly negative) flux error values and makes them NaN.  We might already have such a function.  

For now, I will return to the original behavior of allow zero standard deviation in binned lightcurves to allow the heritage tests to pass, matching the previous expected behavior.

**Comment by gully** [comment]
Great!  The tests are failing due to unrelated `.search()` failures so this PR is ready for review.

**Comment by gully** [comment]
Also TODO:
-[ ] Make a tutorial showing advanced binning patterns.

**Comment by barentsen** [comment]
@gully Can you rebase on master?  The unit tests should pass when you do -- I fixed a couple of bugs today (cf. #631).

File: lightkurve/lightcurve.py [hunk 2]
```diff
@@ -880,16 +881,24 @@ def remove_outliers(self, sigma=5., sigma_lower=None, sigma_upper=None,
             return self[~outlier_mask], outlier_mask
         return self[~outlier_mask]
 
-    def bin(self, binsize=13, method='mean'):
-        """Bins a lightcurve in blocks of size ``binsize``.
+    def bin(self, binsize=None, bins=None, method='mean'):
+        """Bins a lightcurve in chunks defined by `binsize` or `bins`
```
Comment (barentsen) [inline lightkurve/lightcurve.py]:
```suggestion
        """Bins a lightcurve in chunks defined by `binsize` or `bins`.
```

File: lightkurve/lightcurve.py [hunk 3]
```diff
@@ -880,16 +881,24 @@ def remove_outliers(self, sigma=5., sigma_lower=None, sigma_upper=None,
...
         median (``method='median'``) of the original data.  The default is mean.
 
         Parameters
         ----------
-        binsize : int
-            Number of cadences to include in every bin.
+        binsize : int or None
+            Number of cadences to include in every bin.  The default
+            is 13 if neither `bins` nor `binsize` is assigned.
+        bins : int or list or str (optional)
```
Comment (barentsen) [inline lightkurve/lightcurve.py]:
```suggestion
        bins : int, list of int, or str
```

File: lightkurve/lightcurve.py [hunk 4]
```diff
@@ -913,39 +922,44 @@ def bin(self, binsize=13, method='mean'):
...
+            bin_edges = calculate_bin_edges(bin_by_array, bins=n_bins)
+        else: #bins was assigned
+            bin_by_array = self.time
+            bin_edges = calculate_bin_edges(bin_by_array, bins=bins)
+            n_bins = len(bin_edges)-1
+
+        for attr, bin_function in statistic_mapper.items():
+            values_to_bin = getattr(self, attr)
+            # Override error propagation if flux_err is all NaN
+            if (attr == 'flux_err') & np.all(np.isnan(self.flux_err)):
```
Comment (barentsen) [inline lightkurve/lightcurve.py]: This will be faster:
```suggestion
            if (attr == 'flux_err') & np.any(np.isfinite(self.flux_err)):
```

**Comment by barentsen** [comment]
@gully Can you add a warning if the bin edges passed into `bins` are all smaller or larger than the time range?

e.g.
````
if (np.nanmax(bins) < np.nanmin(time)) or (np.nanmax(time) < np.nanmin(bins)):
    warning: the bin edges do no fall in the time range!
```

File: lightkurve/lightcurve.py [hunk 5]
```diff
@@ -913,39 +922,50 @@ def bin(self, binsize=13, method='mean'):
...
+            n_bins = self.flux.size // binsize
+            bin_by_array = np.arange(len(self.time))
+            bin_edges = calculate_bin_edges(bin_by_array, bins=n_bins)
+        else: #bins was assigned
+            bin_by_array = self.time
+            bin_edges = calculate_bin_edges(bin_by_array, bins=bins)
+            n_bins = len(bin_edges)-1
+
+            if ((np.max(bin_edges) < np.min(self.time)) or
+                (np.nanmax(self.time) < np.nanmin(bin_edges)) ):
```
Comment (barentsen) [inline lightkurve/lightcurve.py]: Codacy Issue found: [visually indented line with same indent as next logical line](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4502630)
